components/playlist/browser/resources/components/playlistsCatalog.tsx

```diff
@@ -0,0 +1,149 @@
+
+  /* Default background for playlist without thumbnail */
+  ${ p => !p.isDefaultPlaylist && !p.thumbnailUrl && css`background-color: ${color.light.container.highlight};`}
+
+  /* Playlist with Thumbnail*/
+  ${ p => !p.isDefaultPlaylist && p.thumbnailUrl && css`
+      background: linear-gradient(0deg, rgba(0, 0, 0, 0.4), rgba(0, 0, 0, 0.4)), url(${ p.thumbnailUrl });
+      background-size: cover;` }
+  border-radius: 8px;
+  padding: 8px;
+
+  position:relative;
+  display:flex;
+  gap: 16px;
+`
+
+const PlayLaterCardOverlay = styled.div<{}>`
+  position: absolute;
+  top: 0;
+  right: 0;
+  content: url(${PlayLaterCardOverlayImage});
+`
+
+const ColoredPlaylistInfo = styled(PlaylistInfo)<{ hasBackground: boolean }>`
+  color: ${ ({hasBackground}) => hasBackground ? color.dark.text.primary : color.light.text.primary };
+`
+
+const PlaylistThumbnailContainer = styled.div<ThumbnailProps>`
+  /* Default Playlist icon */
+  ${ p => p.isDefaultPlaylist && css`background-color: ${color.light.interaction.buttonPrimaryBackground}; color: white;` }
+
+  /* Default icon for playlist without thumbnail */
+  ${ p => !p.isDefaultPlaylist && !p.thumbnailUrl && css`background-color: ${color.light.container.background}; `}
+
+  /* Playlist with Thumbnail*/
+  ${ p => !p.isDefaultPlaylist && !!p.thumbnailUrl && css`
+    background-image: url(${p.thumbnailUrl}); 
+    background-size: contain;
+    background-repeat: no-repeat;
+    background-position: center;` }
+
+  width: 80px;
+  height: 80px;
+
+  display: flex;
+  justify-content: center;
+  align-items: center;
+`
+function PlaylistThumbnail (props : ThumbnailProps) {
+    return (
+      <PlaylistThumbnailContainer {...props}>
+        {
+          (props.isDefaultPlaylist || !props.thumbnailUrl) && 
+          <Icon name={props.isDefaultPlaylist ? "history" : "product-playlist"} />
+        }
+      </PlaylistThumbnailContainer>
+    )
+}
+
+// A card UI for representing a playlist.
+function PlaylistCard ({playlist} : {playlist: PlaylistMojo.Playlist}) {
+  const isDefaultPlaylist = React.useMemo(() => {
+    return playlist.id === "default"
+  }, [playlist])
+
+  const thumbnailUrl = React.useMemo(() => {
+    if (!playlist.items) return undefined 
+
+    return playlist.items?.find(item => item.thumbnailPath?.url)?.thumbnailPath.url
+  }, [playlist, playlist.items])
+
+  const hasBackground = React.useMemo(() => {
+    return isDefaultPlaylist || !!thumbnailUrl
+  }, [isDefaultPlaylist, thumbnailUrl]);
+
+  const totalDuration = React.useMemo(() => {
+    // TODO(sko) Duration value is not correct now.
+    //  * We need to update duration when Playlist player plays a video
+    //  * Check if duration is converted well. the duration value is formatted
+    //    as string based on base::Time.
+    if (!playlist.items?.length) return 0
+
+    return playlist.items.reduce((sum, item) => {
+      return sum + parseInt(item.duration)
+    }, 0)
+  }, [playlist, playlist.items])
```

Comment:
Yeah, my brain was about to melt down when I was thinking of this 😭

I was wondering what will happen if we assign a new array to the playlist. item outside this component?`thisPlaylist = []` Then will the memoization will work well?

Reply:
Brain melting is a pretty normal reaction to hooks :laughing: 

But basically, yeah, it should handle reassignments fine :smile:

The React component is rerendered any time:
1. A parent is rendered (this one isn't obvious)
2. Props are changed
3. Redux props are changed

Then the memoization function will be rerun if anything in the deps array is not `===` to the last value it was run with (or it hasn't been run yet).

It won't be rerun if we use `.push(...)` to add an item (but it already won't be). If the playlist changes, but the items don't (i.e. we assigned these items to a new playlist), it doesn't matter, because the result of the calculation would be the same (as the calculation only depends on the items).

Reply:
> It won't be rerun if we use .push(...) to add an item (but it already won't be).

Thanks. @fallaciousreasoning . Now I got more clear view on it. In that case, I guess we'll have problem when the duration value of an item in the array is updated. I've scanned how others do for this, and they seem to make a kind of deep-comparison hacks . Should I do that?

Reply:
I think the generally accepted way of doing this is to never modify items directly, but to create a copy. For example, if we wanted to update the item with `id: 1234`:

```ts
const index = playlist.items.findIndex(p => p.id === '1234')
const copy = [...playlist.items]
copy[index] = {
    ...playlist.items[index],
    duration: 10000
}

playlist.items = copy
```

Unfortunately, you have to make sure this goes up to the top level of state, so if the top level was a dictionary of `{ [playlistId: string]: Playlist }` you'd need to make sure you replaced that too, so React can check for changed keys easily.


```ts
const newState = {
    ...state,
    playlistIdThatIsBeingChanged: {
             ...state[playlistIdThatIsBeingChanged],
             items: copy // the copy from before
    }
}
setState(newState)
```

The idea with this is that you can tell when anything has changed in `O(1)` time (i.e. via `oldState === state`). Unfortunately it's a bit of a pain :laughing:

@petemill do we have any easy utils for doing this in `brave-core`?

Reply:
>  never modify items directly, but to create a copy

Yeah, sounds fair and easy. I should keep that in mind because I've been trying to avoid copy in C++ 😂

Reply:
Hahahaha yeah, JS land can be an interesting place :laughing: